### PR TITLE
JFG-1087: Mark build Unstable if session decision != OK

### PR DESCRIPTION
**Pull request by anton-antonenko**

### CODE REVIEW

File: src/main/java/com/griddynamics/jagger/jenkins/jaas/plugin/JaggerTestExecutionBuilder.java [hunk 1]
```diff
@@ -223,6 +225,38 @@ private TestExecutionEntity waitTestExecutionFinished(PrintStream logger, Long e
...
+        try {
+            logger.println();
+            logger.println(format("Checking decision for test session with id=%s ... ", sessionId));
+            RequestEntity<?> requestEntity = RequestEntity.get(new URI(evaluatedJaasEndpoint + "/sessions/" + sessionId + "/decision")).build();
+            ResponseEntity<DecisionPerSessionDto> responseEntity = restTemplate.exchange(requestEntity, DecisionPerSessionDto.class);
+            Decision decision = responseEntity.getBody().getDecision();
+            logger.println(format("Decision for session with id=%s is %s", sessionId, decision));
+            if (decision != Decision.OK) {
+                String decisionJson = new ObjectMapper().writeValueAsString(responseEntity.getBody());
+                logger.println("Full session decision:\n" + decisionJson);
```
Comment (dlatnikov) [inline src/main/java/com/griddynamics/jagger/jenkins/jaas/plugin/JaggerTestExecutionBuilder.java:251]: let's log full session decision for OK sessions as well. First details, then "Decision for session with id "

File: src/main/java/com/griddynamics/jagger/jenkins/jaas/plugin/JaggerTestExecutionBuilder.java [hunk 2]
```diff
@@ -223,6 +225,38 @@ private TestExecutionEntity waitTestExecutionFinished(PrintStream logger, Long e
...
+                throw new AbortException(format("Test execution is failed due to session with id=%s finished with decision %s.", sessionId, decision));
+            }
+        } catch (URISyntaxException e) {
+            logger.println();
+            throw new AbortException("Invalid JaaS endpoint URL: " + e.getMessage());
+        } catch (RestClientException ex) {
+            logger.println();
+            throw new AbortException(format("Error occurred while checking decision of Test session with id=%s: %s", sessionId, ex.getMessage()));
+        } catch (JsonProcessingException e) {
+            logger.println("\nError occurred:" + e.getMessage());
```
Comment (dlatnikov) [inline src/main/java/com/griddynamics/jagger/jenkins/jaas/plugin/JaggerTestExecutionBuilder.java]: Error occurred => Error occurred during decisions response parsing

Comment (dlatnikov) [inline src/main/java/com/griddynamics/jagger/jenkins/jaas/plugin/JaggerTestExecutionBuilder.java]: can we set session status to "yellow" state in this case?

File: src/main/java/com/griddynamics/jagger/jenkins/jaas/plugin/JaggerTestExecutionBuilder.java [hunk 3]
```diff
@@ -223,6 +225,38 @@ private TestExecutionEntity waitTestExecutionFinished(PrintStream logger, Long e
...
+        else
+            logger.println("Test execution report can be found by the link " + evaluatedJaasEndpoint + "/report?sessionId=" + executionFinished.getSessionId());
+    }
+
+    private void checkDecision(PrintStream logger, String sessionId, RestTemplate restTemplate) throws AbortException {
+        try {
+            logger.println();
+            logger.println(format("Checking decision for test session with id=%s ... ", sessionId));
+            RequestEntity<?> requestEntity = RequestEntity.get(new URI(evaluatedJaasEndpoint + "/sessions/" + sessionId + "/decision")).build();
+            ResponseEntity<DecisionPerSessionDto> responseEntity = restTemplate.exchange(requestEntity, DecisionPerSessionDto.class);
```
Comment (dlatnikov) [inline src/main/java/com/griddynamics/jagger/jenkins/jaas/plugin/JaggerTestExecutionBuilder.java:243]: what if there will be no decision for this session?

Comment (anton-antonenko) [reply src/main/java/com/griddynamics/jagger/jenkins/jaas/plugin/JaggerTestExecutionBuilder.java:243]: Is it possible?

File: src/main/java/com/griddynamics/jagger/jenkins/jaas/plugin/JaggerTestExecutionBuilder.java (hunk 2 above)
Comment (anton-antonenko) [reply src/main/java/com/griddynamics/jagger/jenkins/jaas/plugin/JaggerTestExecutionBuilder.java]: I don't know, how to do it. Need to investigate it.

File: src/main/java/com/griddynamics/jagger/jenkins/jaas/plugin/JaggerTestExecutionBuilder.java (hunk 3 above)
Comment (dlatnikov) [reply src/main/java/com/griddynamics/jagger/jenkins/jaas/plugin/JaggerTestExecutionBuilder.java:243]: yes

**Review by a-badaev — APPROVED** [review]

File: src/main/java/com/griddynamics/jagger/jenkins/jaas/plugin/JaggerTestExecutionBuilder.java [hunk 4]
```diff
@@ -219,13 +227,54 @@ private TestExecutionEntity waitTestExecutionFinished(PrintStream logger, Long e
...
+            }
+        } catch (URISyntaxException e) {
+            logger.println();
+            throw new AbortException("Invalid JaaS endpoint URL: " + e.getMessage());
+        } catch (RestClientException ex) {
+            build.setResult(Result.UNSTABLE);
+            logger.println();
+            if ((ex instanceof HttpClientErrorException) && ((HttpClientErrorException) ex).getRawStatusCode() == 404) {
+                logger.println(format("Error occurred while checking decision of Test session with id=%s: " +
+                        "Endpoint %s/db/sessions/%s/decision is not found. Seems that property 'jaas.hide.db.access.via.api' is set to false.",
```
Comment (dlatnikov) [inline src/main/java/com/griddynamics/jagger/jenkins/jaas/plugin/JaggerTestExecutionBuilder.java]: decisionURL?

Comment (dlatnikov) [inline src/main/java/com/griddynamics/jagger/jenkins/jaas/plugin/JaggerTestExecutionBuilder.java]: propose not to mention jaas properties here. we will change them in jaas and will forget to update them here

File: src/main/java/com/griddynamics/jagger/jenkins/jaas/plugin/JaggerTestExecutionBuilder.java [hunk 5]
```diff
@@ -219,13 +227,53 @@ private TestExecutionEntity waitTestExecutionFinished(PrintStream logger, Long e
         return execution;
     }
 
-    private void publishReportLink(PrintStream logger, TestExecutionEntity executionFinished) {
-        if (executionFinished.getSessionId() == null)
+    private void publishReportLink(PrintStream logger, AbstractBuild<?, ?> build, TestExecutionEntity executionFinished) {
+        if (executionFinished.getSessionId() == null) {
             logger.println("sessionId is unavailable. Can’t publish link to the test report.");
-        else
+            build.setResult(Result.UNSTABLE);
```
Comment (dlatnikov) [inline src/main/java/com/griddynamics/jagger/jenkins/jaas/plugin/JaggerTestExecutionBuilder.java]: not necessary

Comment (anton-antonenko) [reply src/main/java/com/griddynamics/jagger/jenkins/jaas/plugin/JaggerTestExecutionBuilder.java]: Leave it Successful?

**Review by dlatnikov — APPROVED** [review]

**Pull request merged**